lib/grizzly/zwave/commands/thermostat_setpoint_set.ex

```diff
@@ -0,0 +1,75 @@
+defmodule Grizzly.ZWave.Commands.ThermostatSetpointSet do
+  @moduledoc """
+  This module implements command THERMOSTAT_SETPOINT_SET of the COMMAND_CLASS_THERMOSTAT_SETPOINT command class.
+
+  This command is used to set the target value for a given setpoint type
+
+  Params:
+
+    * `:type` - one of :heating | :cooling | :furnace | :dry_air | :moist_air | :auto_changeover (required)
+
+    * `:scale` - temperature scale is used for the setpoint value, one of :celsius or :fahrenheit (default)
+
+    * `:value` - the value of the setpoint
+
+  """
+
+  @behaviour Grizzly.ZWave.Command
+
+  alias Grizzly.ZWave.{Command, DecodeError}
+  alias Grizzly.ZWave.CommandClasses.ThermostatSetpoint
+
+  @type param :: {:type, ThermostatSetpoint.type()} | {:scale, ThermostatSetpoint.scale()}
+
+  @impl true
+  @spec new([param()]) :: {:ok, Command.t()}
+  def new(params) do
+    command = %Command{
+      name: :thermostat_setpoint_report,
+      command_byte: 0x03,
+      command_class: ThermostatSetpoint,
+      params: params,
+      impl: __MODULE__
+    }
+
+    {:ok, command}
+  end
+
+  @impl true
+  def encode_params(command) do
+    type_byte = Command.param!(command, :type) |> ThermostatSetpoint.encode_type()
+
+    scale_byte =
+      (Command.param(command, :scale) || :fahrenheit) |> ThermostatSetpoint.encode_scale()
```

Comment:
I am in favor of not making the scale param as default or using celsius as the default

1. US is the only place that does not use celsius by default, so if we are going to default we should default to what is most useful to the most amount of people
1. This is kinda like defaulting the target value in the binary switch set, so I am not sure that this value should be defaulted.

Also `Command.param/2` will return `nil` by default, but you can pass the default, but you can use `Command.param/3` and pass a default:

```elixir
Command.param(command, :scale, :celsius)
```

Also if we are going to default I do prefer to handle merging the defaults in the `new` function for example: https://github.com/smartrent/grizzly/blob/master/lib/grizzly/zwave/commands/door_lock_operation_report.ex#L90.

The main reason for this is in the logging we when log the outgoing messages if we default in the `encode_params` callback then the log will look like there are missing params, and the defaults are not explicit. This makes reading logs and debugging from those logs a  little confusing. I only say this because I have deafulted in the `encode_params` and wasted a fair amount of time trying to figure out why the log was missing params in the command that I was sending when I knew they should be there.

I have tried to log before encoding happens because otherwise we just get a bunch of bytes and that is much slower to parse mentally then logging the `%Command{}` that is being sent.

Reply:
I agree